### PR TITLE
[ICDS] Added pdf ISSNIP Monthly report and feature flag for new features

### PR DESCRIPTION
**Pull request by lwyszomi**

### CODE REVIEW

File: custom/icds_reports/models.py [hunk 1]
```diff
@@ -15,7 +15,7 @@ class AwcLocation(models.Model):
     district_id = models.TextField()
     district_name = models.TextField(blank=True, null=True)
     district_site_code = models.TextField(blank=True, null=True)
-    state_id = models.TextField()
+    state_id = models.TextField(primary_key=True)
```
Comment (emord) [inline custom/icds_reports/models.py]: Why did you change this primary key? I don't think that's correct

File: custom/icds_reports/models.py [hunk 2]
```diff
@@ -31,7 +31,7 @@ class Meta:
 
 
 class AggAwcDailyView(models.Model):
-    awc_id = models.TextField(primary_key=True)
+    awc_id = models.TextField(blank=True, null=True)
```
Comment (emord) [inline custom/icds_reports/models.py]: I don't htink these can be blank or null. have you seen that?

File: custom/icds_reports/models.py (hunk 1 above)
Comment (lwyszomi) [reply custom/icds_reports/models.py]: I generated models by ```manage.py inspectdb```

File: custom/icds_reports/models.py (hunk 2 above)
Comment (lwyszomi) [reply custom/icds_reports/models.py]: I generated models by ```manage.py inspectdb```

**Comment by fyniac** [comment]
@czue can you please review?  When done, we need to deploy India

**Comment by czue** [comment]
@fyniac sure - just as a quick heads up it looks like this has merge conflicts which should be resolved. @lwyszomi are you able to do that? I can also handle if needed

**Comment by lwyszomi** [comment]
@czue branch updated. ready for review

File: custom/icds_reports/templatetags/icds_format.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        if where_housed == 3:
+            return provided_buildings[building]
+        else:
+            return housed[where_housed]
+    else:
+        return ''
+
+
+@register.filter(name='icds_yes_no')
+def icds_yes_no(value):
```
Comment (czue) [inline custom/icds_reports/templatetags/icds_format.py]: any reason you can't just use django's [`yesno`](https://docs.djangoproject.com/en/2.0/ref/templates/builtins/#yesno)? it takes arguments if you need to customize the text.

File: custom/icds_reports/utils.py [hunk 4]
```diff
@@ -422,3 +425,29 @@ def chosen_filters_to_labels(config, default_interval=''):
...
+    template = get_template("icds_reports/icds_app/pdf/issnip_monthly_register.html")
+    file_name = "{}.pdf".format(file_name)
+    resultFile = open(os.path.join(directory, file_name), "w+b")
+    pisa.CreatePDF(
+        template.render(pdf_context),
+        dest=resultFile)
+    resultFile.close()
+
+
+def have_access_to_features(user):
```
Comment (czue) [inline custom/icds_reports/utils.py]: nit: but I think there could be a better name for this. like `icds_pre_release_features` or something.

File: custom/icds_reports/views.py [hunk 5]
```diff
@@ -642,6 +668,19 @@ def post(self, request, *args, **kwargs):
                 loc_level=aggregation_level,
                 show_test=include_test
             ).to_export('csv', location)
+        elif indicator == 7:
```
Comment (czue) [inline custom/icds_reports/views.py]: probably not worth addressing wholesale, but just noting more magic numbers here

File: custom/icds_reports/views.py [hunk 6]
```diff
@@ -1566,3 +1605,23 @@ class ICDSBugReportView(BugReportView):
...
     def recipients(self):
         return [ICDS_SUPPORT_EMAIL]
+
+
+@method_decorator([login_and_domain_required], name='dispatch')
+class DownloadPDFReport(View):
+    def get(self, request, *args, **kwargs):
+        uuid = self.request.GET.get('uuid', None)
+        format = self.request.GET.get('format', None)
+        pdf_dir = os.path.join(settings.BASE_DIR, 'custom/icds_reports/static/media/')
```
Comment (czue) [inline custom/icds_reports/views.py]: as mentioned already above, this won't work. we'll need to use a storage that isn't tied to the machine/os.

File: settings.py [hunk 7]
```diff
@@ -200,6 +200,8 @@
     'ws4redis',
     'statici18n',
     'raven.contrib.django.raven_compat',
+    'html5lib',
+    'xhtml2pdf',
```
Comment (czue) [inline settings.py]: why do these need to be apps?

**Review by czue — APPROVED** [review]
@lwyszomi this looks like it must have been quite the journey to implement! :clap: 

I think we'll need to rework the PDF file handling, but the rest is mostly questions or small optional comments.

**Review by czue — CHANGES_REQUESTED** [review]
sorry I pressed the wrong option. this needs changes

File: custom/icds_reports/models.py [hunk 8]
```diff
@@ -3,7 +3,7 @@
 
 
 class AwcLocation(models.Model):
-    doc_id = models.TextField(primary_key=True)
+    doc_id = models.TextField()
```
[the remark this answers is not in the document]
Comment (lwyszomi) [reply custom/icds_reports/models.py]: I generated this by ```manage.py inspectdb``` and forgot to restore

File: custom/icds_reports/templatetags/icds_format.py (hunk 3 above)
Comment (lwyszomi) [reply custom/icds_reports/templatetags/icds_format.py]: I didn't know that django have already this functionality :)

**Comment by lwyszomi** [comment]
@czue I forgot that the paths will not work on the production. I refactored code, now I create file in memory and result keep in the cache

Please let me know if you have any additional questions.

**Review by czue — APPROVED** [review]
@lwyszomi looks good. I'm heading offline now but I think this is ok to go from my side once tests run
